### PR TITLE
[Interop][SwiftToCxx] Implement enum case switching

### PR DESCRIPTION
**Pull request by WANGJIEKE**
This is a follow-up of pull request https://github.com/apple/swift/pull/59669. This pull request implements `operator cases()` and various predicates.

For example, following Swift enum

```swift
public enum Foo { case x, y }
```

will have member functions in the generated C++ header file like this

```c++
class Foo final {
public:
  enum class cases {
    x,
    y
  };
  operator cases() const {
    switch (_getEnumTag()) {
      case 0: return cases::x;
      case 1: default: return cases::y;
    }
  }
  bool isX() const {
    return *this == cases::x;
  }
  bool isY() const {
    return *this == cases::y;
  }
};
```

Other changes:

- Print `EnumValueWitnessTable` struct when printing C++ interop scaffold;
- Add `_getEnumTag` helper function for retrieving the tag from enum value witness table.

### CODE REVIEW

**Comment by WANGJIEKE** [comment]
@hyp Could you please review this PR? Thanks!

I have a question regarding naming predicate functions: is `std::toupper` a good way to capitalize a letter? I'm worrying if there will be some corner cases I'm unaware of.

File: test/Interop/SwiftToCxx/enums/enum-cxx-clike-enum-case-functions-execution.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+// RUN: %target-swift-frontend %S/enum-cxx-clike-enum-case-functions.swift -typecheck -module-name Enums -clang-header-expose-public-decls -emit-clang-header-path %t/enums.h
+
+// RUN: %target-interop-build-clangxx -c %s -I %t -o %t/swift-enums-execution.o
+// RUN: %target-interop-build-swift %S/enum-cxx-clike-enum-case-functions.swift -o %t/swift-enums-execution -Xlinker %t/swift-enums-execution.o -module-name Enums -Xfrontend -entry-point-function-name -Xfrontend swiftMain
+
+// RUN: %target-codesign %t/swift-enums-execution
+// RUN: %target-run %t/swift-enums-execution
+
+// REQUIRES: executable_test
```
Comment (hyp) [inline test/Interop/SwiftToCxx/enums/enum-cxx-clike-enum-case-functions-execution.cpp:11]: Please also disable it on arm64e : ‘// UNSUPPORTED: CPU=arm64e’

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 2]
```diff
@@ -384,30 +384,73 @@ class DeclAndTypePrinter::Implementation
...
+      }
+      os << "  };\n"; // enum class cases' closing bracket
+
+      if (elements.size() == 0) {
+        os << "\n";
+        return;
+      }
+
+      // Rearrange cases so that they match their tag values
+      llvm::stable_sort(elements, [&](const auto *e1, const auto *e2) {
```
Comment (hyp) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:413]: How did you determine that this is the correct ordering? Is the compiler doing this kind of sort anywhere else? We might want to reuse this if thins is already being done elsewhere

**Comment by hyp** [comment]
Can you also add check lines for implementation of ‘_getEnumTag’

**Comment by hyp** [comment]
> @hyp Could you please review this PR? Thanks!
> 
> I have a question regarding naming predicate functions: is `std::toupper` a good way to capitalize a letter? I'm worrying if there will be some corner cases I'm unaware of.

For now it’s fine, we already use it for properties. It might not handle Unicode correctly, but we will refactor and fix name transformations later.

File: test/Interop/SwiftToCxx/enums/enum-cxx-clike-enum-case-functions-execution.cpp (hunk 1 above)
Comment (WANGJIEKE) [reply test/Interop/SwiftToCxx/enums/enum-cxx-clike-enum-case-functions-execution.cpp:11]: Oh yeah I forgot this.

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 3]
```diff
@@ -384,30 +384,73 @@ class DeclAndTypePrinter::Implementation
...
+      });
+
+      // Printing operator cases()
+      os << "  inline operator cases() const {\n";
+      os << "    switch (_getEnumTag()) {\n";
+      for (unsigned i = 0; i < elements.size(); ++i) {
+        if (i != elements.size() - 1) {
+          os << "      case " << i << ": return cases::";
+        } else {
+          os << "      case " << i << ": default: return cases::";
```
[the remark this answers is not in the document]
Comment (WANGJIEKE) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp:424]: I want to silence the non-exhaustive switch warning from the compiler.

File: lib/PrintAsClang/DeclAndTypePrinter.cpp (hunk 2 above)
Comment (WANGJIEKE) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp:413]: I got it from the memory layout doc, but yeah it's better to reuse existing functionality. I'll ask this in the office hour meeting.

**Comment by egorzhdan** [comment]
@swift-ci please test

File: lib/PrintAsClang/DeclAndTypePrinter.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (hyp) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp]: For sure. Unfortunately I don’t think the default case should map to a specific case. I think by default it should abort as that’s an unexpected path for fixed enums. Resilient enums are more complicated as they could have an unknown case, but we can handle them in a follow up.

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
Looks good apart from one request

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Comment by WANGJIEKE** [comment]
Also change the return type of `IRABIDetailsProvider::getEnumTagMapping` from `std::map` to `llvm::MapVector` so that it actually returns the cases in their declaration order.

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Review by hyp — APPROVED** [review]
LGTM
